experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,140 @@
+    // Determines the original order in which messages arrived.
+    number: usize,
+    // Arbitrary message data.
+    data: Vec<u8>,
+    // Channel for sending responses back to the client async tasks.
+    response_sender: oneshot::Sender<Response>,
+}
+
+#[async_trait]
+pub trait RequestHandler: Send + Sync {
+    async fn handle(&self, request: Vec<u8>) -> Vec<u8>;
+}
+
+// Trusted Shuffler implementation.
+pub struct TrustedShuffler {
+    // Value k that represents k-anonymity.
+    anonymity_value: usize,
+
+    // Current batch of requests to be shuffled.
+    // Mutex is used because messages are collected from different async tasks.
+    requests_to_shuffle: Arc<Mutex<Vec<Message>>>,
+
+    // Async function that sends a request to the backend and returns a response.
+    request_handler: Arc<dyn RequestHandler>,
+}
+
+impl TrustedShuffler {
+    pub fn new(anonymity_value: usize, request_handler: Arc<dyn RequestHandler>) -> Self {
+        Self {
+            anonymity_value,
+            requests_to_shuffle: Arc::new(Mutex::new(vec![])),
+            request_handler,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (response_sender, response_receiver) = oneshot::channel();
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut requests_to_shuffle = self
+                .requests_to_shuffle
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+
+            let request = Message {
+                number: requests_to_shuffle.len(),
+                data: request,
+                response_sender,
+            };
+
+            requests_to_shuffle.push(request);
+
+            if requests_to_shuffle.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(requests_to_shuffle.deref_mut(), vec![]);
+
+                // Shuffle requests and spawn async tasks for sending them to the backend.
+                let request_handler = self.request_handler.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_handler).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        response_receiver.await.context("Couldn't receive response")
+    }
+
+    // Lexicographically sorts requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(
+        mut requests: Vec<Message>,
+        request_handler: Arc<dyn RequestHandler>,
+    ) -> anyhow::Result<()> {
+        requests.sort_by_key(|v| v.data.clone());
```

Comment:
Turns out that to avoid the cloning, the previous version is necessary: https://stackoverflow.com/a/47126516 . I think we should do that, since in general the requests may be large, and we don't want to clone them unnecessarily.